grapher/stackedCharts/MarimekkoChart.tsx

```diff
@@ -0,0 +1,1539 @@
+                        : this.renderBar(
+                              {
+                                  kind: BarShape.BarPlaceholder,
+                                  seriesName: entityId,
+                              },
+                              {
+                                  ...tooltipProps,
+                                  highlightedSeriesName: "",
+                              },
+                              barWidth,
+                              isHovered,
+                              isSelected,
+                              isFaint,
+                              entityColor?.color
+                          )}
+                </g>
+            )
+            if (isSelected || isHovered) highlightedElements.push(result)
+            else normalElements.push(result)
+        }
+
+        return _.concat(
+            noDataAreaElement ? [noDataAreaElement] : [],
+            normalElements,
+            placedLabels,
+            labelLines,
+            highlightedElements,
+            noDataLabel ? [noDataLabel] : []
+        )
+    }
+    private paddingInPixels = 5
+
+    private renderBar(
+        bar: BarOrPlaceholder,
+        tooltipProps: TooltipProps,
+        barWidth: number,
+        isHovered: boolean,
+        isSelected: boolean,
+        isFaint: boolean,
+        entityColor: string | undefined
+    ): JSX.Element {
+        const { dualAxis, manager } = this
+        const { seriesName } = bar
+        const isPlaceholder = bar.kind === BarShape.BarPlaceholder
+        const barBaseColor =
+            entityColor ?? (bar.kind === BarShape.Bar ? bar.color : "#555")
+
+        const barColor =
+            bar.kind === BarShape.BarPlaceholder
+                ? "#555"
+                : isHovered
+                ? color(barBaseColor)?.brighter(0.9).toString() ?? barBaseColor
+                : isSelected
+                ? color(barBaseColor)?.brighter(0.6).toString() ?? barBaseColor
+                : barBaseColor
+        const strokeColor =
+            isHovered || isSelected ? "#555" : isPlaceholder ? "#aaa" : "#666"
+        const strokeWidth = isHovered || isSelected ? "1px" : "0.5px"
+
+        let barY: number = 0
+        let barHeight: number = 0
+        if (bar.kind === BarShape.Bar) {
+            barY = dualAxis.verticalAxis.place(this.y0 + bar.yPoint.valueOffset)
+            barHeight =
+                dualAxis.verticalAxis.place(this.y0) -
+                dualAxis.verticalAxis.place(bar.yPoint.value)
+        } else {
+            barY = dualAxis.verticalAxis.place(this.y0)
+            barHeight = dualAxis.verticalAxis.rangeSize
+        }
+        const barX = 0
+
+        return (
+            <TippyIfInteractive
+                lazy
+                isInteractive={!manager.isExportingtoSvgOrPng}
+                key={seriesName}
+                hideOnClick={false}
+                content={<MarimekkoChart.Tooltip {...tooltipProps} />}
+            >
```

Comment:
The tooltip is quite slow on some machines:


https://user-images.githubusercontent.com/1308115/127117917-8059b9c1-6a8e-4c83-936d-b9389d2b2051.mov

It seems ~330ms are spent in React rendering + reconciliation on mouseover, probably React is reconstructing a big chunk of the chart on hover, despite not that much changing.

I think the stacked bar also does tooltips this way. Other charts (LineChart, Map) do it more efficiently by rendering the tooltip in a separate tree, outside the chart elements, making it possible to skip a lot of work for React. But that tooltip isn't as good as this one in terms of positioninig.

I think in v1 this is fine, but we need to come up with a better tooltip strategy. Since we know the bounds of everything we render, we could have a chart property `tooltipProps` which includes the coordinates and content of the tooltip, and then it renders outside of the chart elements.

Reply:
Is the tooltip in stacked bar charts also slow for you?

I think you're onto something here:
> probably React is reconstructing a big chunk of the chart on hover, despite no elements except the tooltip changing.

In the Marimekko charts all bars are rendered through a `renderBars` call (which is quite computationally heavy), whereas in stacked bar charts I use React components for the individual bars such that React can render it more efficiently. This could improve the tooltip performance on "some machines".

Reply:
Yes I have it on my todo list to break up the render function and make individual bars into tiny components. For the tooltip I will also change how this is done so that they can just be turned on and off